Test dist/ modules against older perls

tonycoz opened the pull request:
Add testing dist/ modules against older perls.

This currently uses the perldocker/perl-testers images rather than the official perl images since perl-testers includes the older versions, while the official perl images don't.

Unfortunately perl-testers doesn't include threaded perls, but it's an improvement over no testing at all.

I currently only test against a small subset of perls, but this is easily changed.

tonycoz committed:
test dist/ modules against various older perls in CI

This currently has one major limitation: the perl-testers images
aren't threaded and don't have threaded alternatives, so the threads,
threads-shared and Thread-Queue dists are effectively untested on the
older perls.

The perl-testers images have a large number of test modules and their
dependencies installed, which means we're testing against the latest
versions of those dependencies (such as parent.pm) rather than against
the version of that dependency that shipped with that perl.

I tried to use the official perl images, but these only include a
small number of recent perls.  The git repository includes DockerFiles
for the older versions, but these aren't published to Docker Hub.

tonycoz committed:
fix warning/error on 5.8 with ambiguous ?...? operator

Warning: Use of "defined" without parentheses is ambiguous at t/cwd_enoent.t line 32.
Search pattern not terminated at t/cwd_enoent.t line 32.

tonycoz committed:
version loads warnings, which loads Carp in older versions

Carp calls rand().

tonycoz committed:
SelfLoader: /r isn't available for old perls

tonycoz committed:
XSLoader.t: fix local skip() implementation

It failed to output:

  ok N # skip ....

lines for each skipped test, which meant the tests output didn't match
the test count in the plan.

Leont commented:
This is a good idea

jkeenan commented:
Is the new program written in such a way that it could be called in environments other than GH continuous integration (assuming appropriate older `perl` executables and libraries were present)?

khwilliamson commented:
On 8/9/22 18:14, Tony Cook wrote:
> Add testing dist/ modules against older perls.
> 
> This currently uses the perldocker/perl-testers images rather than the 
> official perl images since perl-testers includes the older versions, 
> while the official perl images don't.
> 
> Unfortunately perl-testers doesn't include threaded perls, but it's an 
> improvement over no testing at all.
> 
> I currently only test against a small subset of perls, but this is 
> easily changed.
> 

Thanks to @pali, Devel::PPPort is tested back to 5.6 something in CI.  I 
b elieve it includes threaded.

tonycoz commented:
> Is the new program written in such a way that it could be called in environments other than GH continuous integration (assuming appropriate older `perl` executables and libraries were present)?

I think so.

It doesn't try to install dependencies so they would need to be installed first.  I know at least one module depends on `parent` which isn't a `dist/` module, but that's already installed in the `perl-testers` images.

tonycoz commented:
> For the " XSLoader.t: fix local skip() implementation" commit: can you specify in the commit message what was wrong and what it fixed?

I've updated the commit message for that commit and for the `test dist/ modules against various older perls in CI` commit.

bram-perl commented on Porting/test-dist-modules.pl:
```diff
@@ -0,0 +1,1087 @@
+-f $pppfile
+  or die "No ppport.h found in $pppdir\n";
+
+opendir my $distdir, "dist" or die;
```
It would be better to include `$!` in the die message..
(It's very unlikely to fail but if it does fail then knowing why will be very useful)

Leont replied:
Agreed

bram-perl commented on Porting/test-dist-modules.pl:
```diff
@@ -0,0 +1,1087 @@
+  or die "No ppport.h found in $pppdir\n";
+
+opendir my $distdir, "dist" or die;
+my @dists = sort { lc $a cmp lc $b } grep { /^\w/ && $_ ne "Devel-PPPort" } readdir $distdir;
```
Just an observation (feel free to leave the code as is): skipping `Devel-PPPort` is done in this code;
`version` and `Safe` are excluded later;
That these are treated differently looks a bit inconsistent (but as said I'm fine with leaving it as-is)
One final observation: it doesn't include the reason on why `Devel-PPPort` is excluded, consider adding a comment explaining the why? (The why does become obvious when fully reading the code, it was already tested on line 19)

bram-perl commented on Porting/test-dist-modules.pl:
```diff
@@ -0,0 +1,1087 @@
+
+    print "*** Testing $name ***\n";
+    my $dir = tempdir( CLEANUP => 1);
+    system qq(cp -a "dist/$name/." "$dir/.")
```
why `system EXPR` and not system LIST`? i.e.

    system("cp", "-a", "dist/$name/.",  "$dir/.");

(unlikely to matter tho since `$name`/`$dir` is unlikely to contain special characters that would be treated differently by the shell)

bram-perl commented on Porting/test-dist-modules.pl:
```diff
@@ -0,0 +1,1087 @@
+    print "*** Testing $name ***\n";
+    my $dir = tempdir( CLEANUP => 1);
+    system qq(cp -a "dist/$name/." "$dir/.")
+      and die "Cannot copy files";
```
consider including `$!` and/or `$?`

Leont replied:
I don't really think either would help in this case. $! should be empty, and $? will just contain the return value of the `cp`, which isn't informative at all. `cp` will probably print an error message itself that is helpful.

bram-perl replied:
In my experience that really depends on what happened..
If `cp` encountered an error then yes it will print an informative message; but if something killed the cp command then it will not print a message (and in that case having `$?` in the error is useful since then it's clear that it was killed and with what signal)

(but I'm also ok with leaving as is if it's considered not to be worth it)

bram-perl commented on Porting/test-dist-modules.pl:
```diff
@@ -0,0 +1,1087 @@
+      or die "Cannot chdir $dir: $!";
+    if ($pppfile) {
+        system "cp", $pppfile, "."
+          and die "Cannot copy $pppfile to .\n";
```
consider including `$!` and/or `$?`

bram-perl commented on Porting/test-dist-modules.pl:
```diff
@@ -0,0 +1,1087 @@
+    system "make test TEST_VERBOSE=$verbose"
+      and die "$name: make test failed\n";
+
+    system "make install"
```
Should it include a PREFIX on where to install?
Unless I'm missing something it will be installing in the the default location and not inside a tmpdir?
(That is: the make install may/will change the running system and could - in theory - impact later tests?)
(Also: writing to the default location might require root permissions?)

tonycoz replied:
They're installed so later tests are impacted.

This was written for use within the docker images which are setup to allow installation with the default permissions.

I can see this script being extended to allow use outside of that environment (and allowing invocation like `someperl Porting/test-dist-modules.pl Storable` rather than populating from `readdir()`) but I don't aim to do that in this PR.

bram-perl replied:
Ok, fair enough

bram-perl commented on Porting/test-dist-modules.pl:
```diff
@@ -0,0 +1,1087 @@
+    $dir;
+}
+
+sub write_testpl {
```
This needs a bigger explanation of why this is needed/done..

I.e. something like:

    Some dists (IO, threads, threads-shared) include a t/test.pl file when they're released on CPAN.
    In the repo/in core however there is no t/test.pl in the module dir (i.e. dist/IO/t/test.pl does not exist) and instead there it uses the t/test.pl.
    To test these tests on an older perl we need to add a 't/test.pl' that is suitable for using in the dist dir.
    (blead/core t/test.pl can't be used since it's using newer features not available on older perls)

I do think this is very messy but I suppose cleaning this up isn't as easy/straight-forward..
(I also slightly wonder how/who/where the t/test.pl is added to the dist when releasing to CPAN..)

tonycoz committed:
apply several of the suggestions from @bram-perl

- add an actual error message with `$!` for an opendir failure
- document why Devel-PPPort is treated specially when filtering
  the readdir() results
- change each system() call to list form
- add some explanation to write_testpl()

Also:
- expand die messages for system() failures
